feat: implement spatial navigation

bzizmo opened the pull request:
## Description
This pull request enhances VideoJS by introducing spatial navigation functionality, allowing users to navigate through focusable components within the player using arrow keys.

Spatial Navigation guide https://github.com/videojs/videojs.com/pull/190

## Specific Changes proposed
- Added SpatialNavigation class to manage spatial navigation logic.
- Implemented getPositions() method in components.js to retrieve component position information.
- Enhanced focus and blur event handling in components.js.
- Integrated spatial navigation initialization in player.js.
- Extended slider navigation capabilities in slider.js.

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [ ] If necessary, more likely in a feature request than a bug fix
  - [ ] Change has been verified in an actual browser (Chrome, Firefox, IE)
  - [ ] Unit Tests updated or fixed
  - [ ] Docs/guides updated
  - [ ] Example created ([starter template on JSBin](https://codepen.io/gkatsev/pen/GwZegv?editors=1000#0))
- [ ] Reviewed by Two Core Contributors

dzianis-dashkevich commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   */
+  handlePlayerBlur(component) {
+    if (component.name() === 'CloseButton') {
+      this.refocusComponent();
```
Is it refocusing when closing the player's modal?

dzianis-dashkevich commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+
+    // Iterate inside of all children components of the player.
+    player.children_.forEach((value) => {
+      if (value.hasOwnProperty('el_')) {
```
Does the player have children that are not components?

mister-ben replied:
There can be children that are DOM elements (the video el), and children that are components but which do not have an `el_` property (MediaLoader)

dzianis-dashkevich replied:
I am curious why MediaLoader is a child of the player... (if it does not represent any UI)

dzianis-dashkevich commented:
Could you please also add an example page to play with the feature?

dzianis-dashkevich commented:
Could you please add some documentation about usage and different use-cases for the feature?

wseymour15 commented on src/js/modal-dialog.js:
```diff
@@ -454,17 +461,25 @@ class ModalDialog extends Component {
 
-    if (keycode.isEventKey(event, 'Escape') && this.closeable()) {
+    // If 'Esc' is pressed or Backspace is pressed & spatialNavigation is enabled & Modal is 'closeable'.
+    if (keycode.isEventKey(event, 'Escape') || (keycode.isEventKey(event, 'Backspace') && isSpatialNavlistening) && this.closeable()) {
```
Should we ensure we only act on the `Escape` keycode here when spatial navigation is NOT enabled? Or do we want to close the modal no matter what if escape is hit?

CarlosVillasenor replied:
Hello, we want to close the modal no matter what if escape is hit, the way this line of code is written is to be not intrusive with the key presses  of the player that where already  working.

wseymour15 commented on src/js/modal-dialog.js:
```diff
@@ -454,17 +461,25 @@ class ModalDialog extends Component {
       this.close();
       return;
+    // If 'Enter' is pressed & spatialNavigation is enabled & handleClick is available.
+    } else if (keycode.isEventKey(event, 'Enter') && isSpatialNavlistening && this.handleClick) {
```
Similar question as the above, but for handleClick

CarlosVillasenor replied:
> Similar question as the above, but for handleClick

In this case we only want to handle click if the spatial-navigation is enabled.

wseymour15 commented on src/js/component.js:
```diff
@@ -1283,6 +1287,94 @@ class Component {
+      }
+
+      if (!(event.currentTarget.contains(event.relatedTarget)) && !isChildrenOfPlayer || !nextFocusedElement) {
+        spatialNavigation.handlePlayerBlur(this);
```
On that note, maybe we can have a check before any of this logic is executed that checks if the component is a player

Or even better, don't bind these spatial navigation specific functions if the component is the player

amtins replied:
I don't quite understand the purpose of this function, because even if it's removed it seems to work correctly. What would be the use case where it's needed?

Same applies to `handleFocus`.

CarlosVillasenor replied:
> I don't quite understand the purpose of this function, because even if it's removed it seems to work correctly. What would be the use case where it's needed?
> 
> Same applies to `handleFocus`.

Hello @amtins ,  A use for  case 'handleBlur' would be: The user presses down when is focus is in the control bar, selects a component below the player & then presses up to regain focus of the last selected component inside the control bar. This is possible because we conditionally call spatialNavigation.handlePlayerBlur()  from handleBlur() to store the last selected component on blur.

dzianis-dashkevich replied:
@CarlosVillasenor, @bzizmo 
Do we need to listen to blur/focus event on each component?
Can we move this logic related to handlePlayerFocus and handlePlayerBlur to the spatialNavigation class and listen to these events on the player itself?

eg:

player.on('focus', this.handlePlayerFocus);
player.on('blur', this.handlePlayerBlur);

bzizmo replied:
@dzianis-dashkevich both listeners were moved to spatialnavigation class

wseymour15 commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    const focusableComponents = [];
+
+    /**
+   * Searchs for children candidates.
```
nit
```suggestion
   * Searches for children candidates.
```

wseymour15 commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   * @return {HTMLElement|null} Returns the suitable child element if found, or null if not found.
+   */
+  findSuitableDOMChild(component) {
+    function searchForSuitableChild(node) {
```
nit: It seems a little unecessary to create an extra function here, you could just do:

```
const node = component.el()

if (component.getIsFocusable(node) && component.getIsAvailableToBeFocused(node)) {
        return node;
}

...

```

wseymour15 commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   * @return {Array}
+   *         Returns an array of focusable components.
+   */
+  getComponents() {
```
I think the naming and documentation of this function is a little confusing when this intention is actually to update `this.focusableComponents`.. maybe it could be renamed to `updateFocusableComponents`..

Then, getComponents could just return `this.this.focusableComponents`

CarlosVillasenor replied:
Made a commit to rename 'focusableComponents' to 'updateFocusableComponents', regarding "getComponents could just return this.this.focusableComponents" I think is a good idea & we can implement it when we required to just return the focusable components.

wseymour15 commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   * @param {string} eventName - The name of the event to notify listeners for.
+   * @param {Object} [eventDetails={}] - Additional details to include in the event object.
+   */
+  notifyListeners(eventName, eventDetails = {}) {
```
Could we take advantage EventTarget class and avoid some of this logic?

We could have the SpatialNavigation class extend `EventTarget`, and we could use the methods in that class to handle events

bzizmo replied:
Thanks for pointing this out @wseymour15 
I've changed SpatialNavigation class to extend EventTarget and removed the redundant methods.

amtins commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   * Starts the spatial navigation by adding a keydown event listener to the video container.
+   * This method ensures that the event listener is added only once.
+   */
+  start() {
```
This function is never called, preventing spatial navigation from working.

bzizmo replied:
Thank you for pointing this out, @amtins. The decision to not automatically start the spatial navigation was intentional, to provide developers with the flexibility to enable it based on their specific application requirements. There are scenarios, such as when using a spatial-navigation-polyfill, where it might be beneficial to have the "videojs spatialNavigation" functionality integrated and available but not active.

For more information on how to implement and configure spatial navigation, including how to manually start it, please refer to our draft documentation here: [Spatial Navigation Guide](https://github.com/bzizmo/videojs.com/blob/main/src/mdx-pages/guides/spatial-navigation.mdx).

amtins commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   *                                                   If null or not provided, no component will be initially focused.
+   */
+  constructor(player, initialFocusedComponent) {
+    this.KEY_CODE = { 37: 'left', 38: 'up', 39: 'right', 40: 'down', 415: 'play', 19: 'pause', 417: 'ff', 412: 'rw' };
```
Wouldn't it be possible to extend `keycode` with the missing keys to avoid having two ways of handling keys?

Let's say somewhere in the `utils` directory something like:
```javascript
import keycode from 'keycode';

keycode.codes.play = 415;
keycode.codes.ff = 417;
keycode.codes.rw = 412;

export keycode;
```

bzizmo replied:
Sure @amtins I can move the keycodes to another file.
@dzianis-dashkevich would be this a viable solution for you?

dzianis-dashkevich replied:
I am fine with this

bzizmo replied:
Changed as per @amtins suggestion in the latest commit.

amtins commented on src/js/tracks/text-track-settings-colors.js:
```diff
@@ -0,0 +1,109 @@
+    this.addChild(ElWinColorFieldset);
+
+    // Add created components
+    this.el().appendChild(ElFgColorFieldset.el());
+    this.el().appendChild(ElBgColorFieldset.el());
+    this.el().appendChild(ElWinColorFieldset.el());
```
There's no need to do this, because `addChild` does it.

amtins commented on src/js/tracks/text-track-settings-colors.js:
```diff
@@ -0,0 +1,109 @@
+ *
+ * @extends Component
+ */
+class TrackSettingsColors extends Component {
```
The file name seems to be inconsistent with the class name.

amtins commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+  constructor(player, initialFocusedComponent) {
+    this.KEY_CODE = { 37: 'left', 38: 'up', 39: 'right', 40: 'down', 415: 'play', 19: 'pause', 417: 'ff', 412: 'rw' };
+    this.player = player;
+    this.components = new Set();
```
this property does not seem to be used anywhere

bzizmo replied:
Removed, because was not used.

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   */
+  constructor(player, initialFocusedComponent) {
+    this.KEY_CODE = { 37: 'left', 38: 'up', 39: 'right', 40: 'down', 415: 'play', 19: 'pause', 417: 'ff', 412: 'rw' };
+    this.player = player;
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    this.player = player;
+    this.components = new Set();
+    this.focusableComponents = [];
+    this.isListening = false;
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    this.components = new Set();
+    this.focusableComponents = [];
+    this.isListening = false;
+    this.isPaused = false;
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    this.focusableComponents = [];
+    this.isListening = false;
+    this.isPaused = false;
+    this.eventListeners = [];
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    this.isListening = false;
+    this.isPaused = false;
+    this.eventListeners = [];
+    this.onKeyDown = this.onKeyDown.bind(this);
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    this.isPaused = false;
+    this.eventListeners = [];
+    this.onKeyDown = this.onKeyDown.bind(this);
+    this.currentFocus = initialFocusedComponent || null;
```
Removed in the latest commit as it wasn't used.

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    this.eventListeners = [];
+    this.onKeyDown = this.onKeyDown.bind(this);
+    this.currentFocus = initialFocusedComponent || null;
+    this.lastFocusedComponent = null;
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    this.currentFocus = initialFocusedComponent || null;
+    this.lastFocusedComponent = null;
+    // The number of seconds the `step*` functions move the timeline.
+    this.STEP_SECONDS = 5;
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   *                                                   when the spatial navigation system starts.
+   *                                                   If null or not provided, no component will be initially focused.
+   */
+  constructor(player, initialFocusedComponent) {
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+      this.player.el().addEventListener('keydown', this.onKeyDown);
+      this.isListening = true;
+      // this set focus is currently here just for testing purposes
+      this.focus(this.getComponents()[0]);
```
Resolved in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   * This method ensures that the event listener is added only once.
+   */
+  start() {
+    if (!this.isListening) {
```
Resolved in the latest commit

bzizmo commented on src/js/player.js:
```diff
@@ -562,6 +563,13 @@ class Player extends Component {
+    // Check if spatial navigation is enabled in the options.
+    // If enabled, instantiate the SpatialNavigation class.
+    if (options.spatialNavigation && options.spatialNavigation.enabled) {
+      this.spatialNavigation = new SpatialNavigation(this, null);
```
Removed in the latest commit

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+    } else {
+      this.pause();
+
+      if (component && component.el_) {
```
Resolved

bzizmo commented on src/js/spatial-navigation.js:
```diff
@@ -0,0 +1,496 @@
+   * @param {KeyboardEvent} e - The keydown event to be handled.
+   */
+  onKeyDown(e) {
+    const key = this.KEY_CODE[e.keyCode];
```
I think we should allow the user perform any MediaAction even though the Spatial Navigation is paused, which means the focus is currently somewhere outside player.

CarlosVillasenor commented on src/js/modal-dialog.js:
```diff
@@ -236,7 +236,10 @@ class ModalDialog extends Component {
+
     const player = this.player();
+    const spatialNavigation = this.player_.spatialNavigation;
+    const isSpatialNavlistening = spatialNavigation && spatialNavigation.isListening && !spatialNavigation.isPaused;
```
Updated, done as suggested.

CarlosVillasenor commented on src/js/tracks/text-track-settings.js:
```diff
@@ -254,12 +262,49 @@ class TextTrackSettings extends ModalDialog {
+
+    this.addChild(trackSettingsFont);
+
+    const trackSettingsControls = new TrackSettingsControls(player, {});
```
Thanks, this is solved now.

bzizmo commented on src/css/components/_control-bar.scss:
We need 1px gap in case app has already implemented spatial-navigation-polyfill

bzizmo commented on src/js/component.js:
```diff
@@ -109,6 +109,10 @@ class Component {
+
+      // Binding event handlers
+      this.on('focus', this.handleFocus.bind(this));
+      this.on('blur', this.handleBlur.bind(this));
```
Fixed in the latest commit

bzizmo commented on src/css/components/_control-bar.scss:
```diff
@@ -6,6 +6,7 @@
   left: 0;
   right: 0;
   height: 3.0em;
+  gap: 1px;
```
Added an exception for gap 1px when spatial navigation is enabled

bzizmo commented on src/js/player.js:
```diff
@@ -5434,6 +5442,10 @@ Player.prototype.options_ = {
   audioPosterMode: false,
+  spatialNavigation: {
+    enabled: false,
+    horizontalSeek: false
```
Yes this is an optional param, set by defaut to false. Would you suggest to remove it in this case?

wseymour15 replied:
I think this is fine as is. I think once documentation is created with what this option is expected to look like, that will make this more clear. Resolving 👍

bzizmo commented:
> Why do we use `this.options_.playerOptions.spatialNavigation` everywhere? Should it be available via just options?

I think it's not available as this is player option and not components.

CarlosVillasenor commented on src/js/control-bar/text-track-controls/caption-settings-menu-item.js:
```diff
@@ -53,6 +53,32 @@ class CaptionSettingsMenuItem extends TextTrackMenuItem {
   handleClick(event) {
     this.player().getChild('textTrackSettings').open();
+    const spatialNavigation = this.player_.spatialNavigation;
+    const isSpatialNavlistening = spatialNavigation && spatialNavigation.isListening && !spatialNavigation.isPaused;
```
Thanks, it is solved right now.

CarlosVillasenor commented:
> Could you please also add an example page to play with the feature?

Sure, here is an index we use inside of the sandbox directory: 
https://github.com/CarlosVillasenor/video.js-sandbox-index/blob/main/basic/index.html

CarlosVillasenor commented on src/js/component.js:
```diff
@@ -1765,6 +1857,160 @@ class Component {
+    if (!(el.tabIndex < 0) && !isExpresslyInert(el) && !isActuallyDisabled(el)) {
+      return true;
+    }
+    return false;
```
Tested this changes, they work good in the spatial-navigation added this changes in the last commit.

amtins commented:
## Environment

Test performed with media containing subtitles on a Tizen web application deployed on Samsung M5 display.

## Test result

Generally speaking, the functionality works well, but some edge cases have been found.

### Edge cases

On Tizen, the `back` key is named `XF86Back` and its code is `10009`. As a result, once the *subtitle menu* is open, it is no longer possible to exit the menu. Pressing the `back` key on the remote control has no effect, unless selecting a menu item to force it to close. Similar behavior can be observed in the *captions settings* menu and the *audio menu*.

The progress bar is positioned above the control buttons:

- When the focus is on the subtitle button and the `up` button on the remote control is pressed, the subtitle menu opens instead of the progress bar.
- When the focus is on the play button, pressing the `up` button on the remote selects the progress bar. However, if the `down` button is pressed on the remote control, the subtitle button gets the focus instead of the play button.


## Observation

I'm sure this comment comes a little late, given the state of progress of the work.

Why is `SpanitialNavigation` not a component?

This would have offered several advantages:

1. Removes the coupling between `SpatialNavigation` and `Component` (all components).
2. `SpatialNavigation` would be in charge of all the logic and therefore easier to replace
3. Takes advantage of the mechanism offered by a component regarding `options`, and therefore offer a more flexible developer experience.

For example today:

1. It is not possible for a developer to extend the `SpatialNavigation` class. This prevents customization of the experience, but more importantly, if a bug is present or a feature is missing, the developer will have no choice but to wait for a new release, and therefore won't be able to develop his own workaround.
2. Once the player has been created, it is no longer possible to activate space navigation.

### Alternative solution

```javascript
videojs.registerComponent(
  'SpatialNagivation',
  class extends videojs.getComponent('Component') {
    constructor(player, options, ready) {
      // Spatial navigation component doesn't have a dom element
      options.createEl = false;

      super(player, options, ready);
    }

    ready() {
      // Allow to remove the logic from Player constructor
      if (this.options().enabled) {
        this.start();
        this.addClass('vjs-spatial-navigation-enabled');
      }

      if (this.options().focusElement) {
        this.focus(this.options().focusElement);
      }
    }

   focus(element){
      const component = this.player().getDescendant(element);

      if(component && this.getIsAvailableToBeFocused (component.el())){/*...*/}
      // ...
   }

   getIsAvailableToBeFocused (el) { // allow to remove the logic from the class Component
    // ...
   }
  }
);

const player1 = videojs('player1');
player1.spatialNavigation; // undefined, since Player.prototype.options_.spatialNavigation is false by default

const player2 = videojs('player2', {
  spatialNavigation: true,
});
player2.spatialNavigation; // spatialNavigation component instance
player2.spatialNavigation.start(); // Manually activated when desired

// start the spacial navigation and select the bigPlayButton
const player3 = videojs('player3', {
  spatialNavigation: {
    enable: true,
    focusElement: 'bigPlayButton'
  },
});

// start the spacial navigation and select the playToggle
const player4 = videojs('player4', {
  bigPlayButton: false
  spatialNavigation: {
    enable: true,
    focusElement: ['controlBar', 'playToggle']
  },
}, () => {
  player.addClass('vjs-has-started');
});

const player5 = videojs('player5');
player5.addChild('spatialNavigation', {
  id: 'spatialNavigation',
  enable: true
});
player5.getChild('spatialNavigation').stop();
player5.getChildById('spatialNavigation').start();
player5.spatialNavigation; // is undefined

// Let's say there is a bug
videojs.registerComponent(
  'FixedSpatialNagivation', class extends videojs.getComponent('SpatialNagivation') {
   focus(element){/*my fix*/}
  }
);

const player6 = videojs('player6',{
  fixedSpatialNagivation: true
});

// Or
videojs.registerComponent(
  'SpatialNagivation', class extends videojs.getComponent('SpatialNagivation') {
   focus(element){/*my fix*/}
  }
);

const player6bis = videojs('player6bis',{
  spatialNagivation: true
});

// Let's say I want to enhance the SpatialNagivation and decide which component should be focusable
videojs.registerComponent(
  'SpatialNagivation', class extends videojs.getComponent('SpatialNagivation') {
    constructor(player,options, ready){
      super(player,options, ready);
    }

    ready(){
      const focusableComponents = this.options().focusableComponents || [];
      focusableComponents.forEach(focusableComponent => {
        const component = findComponent(focusableComponent);

        if(!component) return;

        this.focusableComponents.push(component)
      })

      // do something with keycodes option

    }
  }
);

const player7 = videojs('player7', {
  spatialNagivation: {
    focusableComponents : ['bigPlayButton', 'playToggle'], // no other component is focusable
    keycodes : {
      XF86Back: 10009
    }
  }
});
```

### Conclusion

The idea would be to move the spatial navigation code from `Component` to `SpatialNavigation`. This avoids adding code to `Component`. It also reverses control so that the `SpatialNavigation` class keeps responsibility for orchestrating everything to do with spatial navigation, from component lookup to event management.

dzianis-dashkevich commented:
@amtins Thank you for your feedback!

> It is not possible for a developer to extend the SpatialNavigation class. This prevents customization of the experience, but more importantly, if a bug is present or a feature is missing, the developer will have no choice but to wait for a new release, and therefore won't be able to develop his own workaround.


SpatialNavigation does not represent any element in the dom. It is a service, while a component is an abstraction for a DOM element. If we want to use several aspects of a component’s functionality, we should probably use composition over inheritance...


Regarding SpatialNavigation extension.
I guess we should create an API for that, something like that:

```js
import videojs from 'video.js'

class CustomSpatialNavigation extends videojs.getSpatialNavigationFactory() {}

videojs.setSpatialNavigationFactory(CustomSpatialNavigation)
```

> Once the player has been created, it is no longer possible to activate space navigation.

Could you please elaborate more here? I thought it is possible to call: `player.spatialNavigation.start()`

> It also reverses control so that the SpatialNavigation class keeps responsibility for orchestrating everything to do with spatial navigation, from component lookup to event management.

Based on my understanding, the current implementation of the SpatialNavigation class keeps responsibility for event management and component lookup. Could you please elaborate more on what you think should be updated?

mister-ben commented:
> a component is an abstraction for a DOM element.

FWIW we do have components that have no DOM element - the `MediaLoader` and `LiveTracker` have none, and the `ResizeManager` has an iframe only if needed as fallback where `ResizeObserver` is unavailable.

dzianis-dashkevich commented:
> FWIW we do have components that have no DOM element - the MediaLoader and LiveTracker have none, and the ResizeManager has an iframe only if needed as fallback where ResizeObserver is unavailable.

@mister-ben, tbh, I don't really understand why all of them are components... 
- They are sounds like services to me
- They all should not be included in the components tree (should not be children of the player)

PS: I am not opposed to the idea of having SpatialNavigation as a component. I just don't really understand why a service should be treated as a Component and what are the benefits of using it as a component instead of a service.

I would really appreciate it if you could elaborate more on why it was done that way, so I can understand better the origins...

I see a bunch of issues with fake components:
For example: components look up - should be simple DFS/BFS traverse.
But now we have to check whether it is a "real" component or not.

IMHO:
if we need some functionality (that is not related to `this.el_`) of the component class, we should extract this functionality to a separate class so we can re-use it via composition instead of extending from Component and disabling `el_` functionality.
If we need to provide API for extending services, we could just create a separate API via service locator + factory overrides.

mister-ben commented:
It was the case that "everything" is a component for a [very long time](https://github.com/videojs/video.js/blob/2138d4f2a0607820203cb9cdfbf3952f1fcf3d04/src/js/media.loader.js), and I wasn't there for the original design decision. Components track their child components to `dispose()` them for one thing. We need to separate out DOM and non-DOM stuff more, but that will be a significant change. My point is really that _if_ there is a benefit to using a component for this now (I haven't looked into this one enough to judge) then it the absence of a DOM element needn't preclude that based on how things are today.

amtins commented:
@bzizmo, @CarlosVillasenor  I've tested the latest modifications and I can see that the Tizen experience has improved. ❤️

## Should spacial navigation be a service or a component?

@dzianis-dashkevich  to this question, the most important thing is that a developer can easily modify/enrich the experience. **Your proposal addresses this use case.**

Although I think your proposal is excellent, it introduces a new concept into `video.js`. Is this really desirable?

If I take this a step further and focus solely on the tools provided by `video.js` to enrich the experience. The framework has two very powerful functionalities: `components` and `plugins`.

`Components` allow you to be part of the component tree, not necessarily all components have a DOM as mister-ben mentioned, but the arguments you [mentionned](https://github.com/videojs/video.js/pull/8570#issuecomment-1966983655) are completely valid.

`Plugins` are closer to your proposal, since they offer a notion of `function factory`. For example, `quality-levels` is a plugin included by default in `video.js`.

The advantage of either approach is that they are `perfectly integrated` into the `video.js` ecosystem and benefits from the life cycle offered by `component/plugin`.

## Spatial navigation initialization

> Could you please elaborate more here? I thought it is possible to call: player.spatialNavigation.start()

With the current solution, if spatial navigation is not initialized when the player is instantiated, it is no longer possible to do so. However, `components` or `plugins` allow this.

### Demo

#### Actual

```javascript
const player = videojs('player');

// At this player.spatialNavigation is undefined and can't be instantiated anymore
```

#### Component

Please refer to the [Alternative solution](https://github.com/videojs/video.js/pull/8570#issuecomment-1963040916)

#### Plugin

```javascript
videojs.registerPlugin('spatialNavigation', class extends videojs.getPlugin('plugin') {
  constructor(player, options) {
    super(player, options);
  }
});

const player = videojs('player');
// player.spatialNavigation is a createPluginFactory

player.spatialNavigation({enable:true});
// player.spatialNavigation becomes an instance
player.spatialNavigation();
// can be disposed
player.spatialNavigation().dispose();

// OR
const player2 = videojs('player', {
  plugins: {
    spatialNavigation : {enable:true}
  }
});
```

```javascript
// Let's say spatialNavigation already exists and a developer wants to override it
videojs.registerPlugin('spatialNavigation', class extends videojs.getPlugin('spatialNavigation') {
  constructor(player, options) {
    super(player, options);
    console.log('my custom spatialNavigation', options.custom);
  }
});

const player = videojs('player', {
  plugins: {
    spatialNavigation: {
      enable: true,
      custom: 'My custom option'
    }
  }
});
```

## Does Component really have to be changed?

> Based on my understanding, the current implementation of the SpatialNavigation class keeps responsibility for event management and component lookup. Could you please elaborate more on what you think should be updated?

Today's implementation adds a number of functions that don't need to be in `Component`: `getPositions`, `getIsDisabled`, `getIsExpresslyInert`, `getIsFocusable`, `getIsAvailableToBeFocused`

These functions only retrieve information relating to the DOM. They could therefore be added to the DOM utils, or become SpatialNavigation static functions or added somewhere else. Apart from spatial navigation, they have no other use for a component.

*I would also have been inclined not to have internal functions to facilitate unit test writing.*

## Miscellaneous

The word `pause` can be confusing in the context of a player. Wouldn't `suspend` be a more appropriate antonym?

The `lastFocusedComponent_` property never seems to be updated.

When an error occurs, pressing the ESC key closes the error modal.

When the `player.src` changes, the focus is lost.

Scroll is not supported. For example, if text track settings display a scroll bar, only the elements displayed are focusable.

bzizmo commented:
@amtins Thanks for your feedback 
Re Miscellaneous:

> The word `pause` can be confusing in the context of a player. Wouldn't `suspend` be a more appropriate antonym?

This was already discussed and approved in RFCS https://github.com/videojs/rfcs/blob/main/proposals/0002-spatial-navigation.md

> The `lastFocusedComponent_` property never seems to be updated.

This property is updated when a user moves the focus outside the player

> When an error occurs, pressing the ESC key closes the error modal.

Fixed in the latest commit.

> When the `player.src` changes, the focus is lost.

Fixed in the latest commit.

> Scroll is not supported. For example, if text track settings display a scroll bar, only the elements displayed are focusable.

Fixed in the latest commit.

wseymour15 commented on src/js/tracks/text-track-fieldset.js:
```diff
@@ -0,0 +1,129 @@
+   *        A text with part of an string to create atribute of aria-labelledby.
+   *        It passes to 'TextTrackSelect'.
+   *
+   * @param {string} [options.legendText]
```
I'd double check that the documentation for these new track related components are accurate. For example, I don't think we need `options.legendText` here since we aren't actually ever using legendText passed through that `options` object. I think this is the case for a handful of places in these components.

CarlosVillasenor replied:
Hello @wseymour15  legend Text is used to create inside the component like this: 

```
const legendElement = Dom.createEl('legend', {
      textContent: this.localize(this.options_.legendText),
      id: this.options_.legendId
});
```

Is also used in every new instance of the component
<img width="406" alt="Screenshot 2024-04-16 at 10 34 51" src="https://github.com/videojs/video.js/assets/21351830/e1f08c28-183b-446b-8fd9-d27717d9e694">

wseymour15 replied:
Fair enough, I guess it is truly used in a component that this extends, but you are probably right that it is fine to leave here if it is common usage.

bzizmo commented on docs/examples/react-video-nav-app/package.json:
```diff
@@ -0,0 +1,39 @@
+{
```
react app was removed from sandbox

CarlosVillasenor commented on src/js/control-bar/text-track-controls/caption-settings-menu-item.js:
```diff
@@ -53,6 +53,33 @@ class CaptionSettingsMenuItem extends TextTrackMenuItem {
+
+    if (componentToFocus) {
+      spatialNavigation.focus(componentToFocus);
+    }
```
Thanks @wseymour15 this was implemented in the last PR we merged.

CarlosVillasenor commented on src/js/control-bar/text-track-controls/caption-settings-menu-item.js:
```diff
@@ -53,6 +53,33 @@ class CaptionSettingsMenuItem extends TextTrackMenuItem {
+   * This gets called by 'handleClick' if 'spatialNavigation' is enabled.
+   * Searches for the first 'TextTrackSelect' inside of modal to focus.
+   */
+  searchForTrackSelect() {
```
@wseymour15 We moved this function to the spatial navigation in the last merged PR for the branch.

wseymour15 replied:
Yeah, I still think this can be private if it only used in this class.

CarlosVillasenor replied:
> Yeah, I still think this can be private if it only used in this class.

@wseymour15 thanks for the approval I think I will merge the PR so we do not lose the approvals & do an small PR with that change

CarlosVillasenor commented on src/js/component.js:
```diff
@@ -1765,6 +1859,154 @@ class Component {
+    }
+
+    // If element is visible, is being rendered, does not have a parent element & its tabIndex is not negative.
+    if (isVisible(el) && isBeingRendered(el) && ((!el.parentElement) || (el.tabIndex >= 0))) {
```
Thanks @wseymour15 , I updated the comment to: 
// If element is visible, is being rendered & either does not have a parent element or its tabIndex is not negative.

wseymour15 reviewed (approved):
+1, great work! One nit comment left from me.
